test/core/pipeline/test_async_pipeline.py

```diff
@@ -0,0 +1,26 @@
+import asyncio
+
+from haystack import AsyncPipeline
+
+
+def test_async_pipeline_reentrance(waiting_component, spying_tracer):
+    pp = AsyncPipeline()
+    pp.add_component("wait", waiting_component())
+
+    run_data = [{"wait_for": 1}, {"wait_for": 2}]
+
+    async_loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(async_loop)
+
+    async def run_all():
+        # Create concurrent tasks for each pipeline run
+        tasks = [pp.run_async(data) for data in run_data]
+        await asyncio.gather(*tasks)
+
+    try:
+        async_loop.run_until_complete(run_all())
+        component_spans = [sp for sp in spying_tracer.spans if sp.operation_name == "haystack.component.run_async"]
+        for span in component_spans:
+            assert span.tags["haystack.component.visits"] == 1
+    finally:
+        async_loop.close()
```

Comment:
Something like this, although I didnt test it?

```suggestion
def test_async_pipeline_reentrance(waiting_component, spying_tracer):
    """
    Test that the AsyncPipeline can execute multiple runs concurrently and that
    each component is called exactly once per run (as indicated by the 'visits' tag).
    """
    async_pipeline = AsyncPipeline()
    async_pipeline.add_component("wait", waiting_component())

    run_data = [{"wait_for": 1}, {"wait_for": 2}]

    async def run_all():
        tasks = [async_pipeline.run_async(data) for data in run_data]
        await asyncio.gather(*tasks)

        component_spans = [
            sp for sp in spying_tracer.spans
            if sp.operation_name == "haystack.component.run_async"
        ]
        for span in component_spans:
            expected_visits = 1
            actual_visits = span.tags.get("haystack.component.visits")
            assert actual_visits == expected_visits, (
                f"Expected {expected_visits} visit, got {actual_visits} for span {span}"
            )

    # Use asyncio.run to manage the event loop.
    asyncio.run(run_all())
```